ts4.1/index.d.ts

```diff
@@ -111,12 +115,35 @@ type NormalizeMulti<T, U extends keyof T, L = LastOf<U>> = L extends U
   ? AppendNS<L, Normalize<T[L]>> | NormalizeMulti<T, Exclude<U, L>>
   : never;
 
+type CustomTypeParameters = {
+  returnNull: boolean;
+  returnEmptyString: boolean;
+};
+
+type NonNullableTranslation<
+  T,
+  U extends Partial<CustomTypeParameters>
+> = U['returnNull'] extends false ? (T extends null ? string : T) : T;
+
+type NonEmptyStringTranslation<
+  T,
+  U extends Partial<CustomTypeParameters>
+> = U['returnEmptyString'] extends false ? (T extends '' ? string : T) : T;
+
+/**
+ * Checks if user has enabled `returnEmptyString` and `returnNull` options to retrieve correct values.
+ */
+export type NormalizeByTypeOptions<
+  T,
+  U extends Partial<CustomTypeParameters> = TypeOptions
+> = NonNullableTranslation<NonEmptyStringTranslation<T, U>, U>;
```

Comment:
As both helpers are pretty much the same, you could make it more reusable, something like:
```suggestion
type TypeOptionsFallback<T, U, R> = U extends false ? (T extends R ? string : T) : T

/**
 * Checks if user has enabled `returnEmptyString` and `returnNull` options to retrieve correct values.
 */
export type NormalizeByTypeOptions<
  T,
  U extends Partial<CustomTypeParameters> = TypeOptions,
  R = TypeOptionsFallback<T, U['returnEmptyString'], ''>
> = TypeOptionsFallback<R, U['returnNull'], null>;
```

Reply:
Yup! I applied your suggestion, but changing the generic parameters naming. What do you think ? @pedrodurek

Reply:
Sounds good to me!